src/adapters/env.ts

```diff
@@ -0,0 +1,110 @@
+import { z, ZodError } from "zod";
+
+import { StrictSchema } from "src/utils/type-safety";
+import { ZodEither } from "src/utils/types";
+// adapters
+import * as adapters from "src/adapters";
+
+interface ProductionLiteral {
+  REACT_APP_ENV: "production";
+}
+
+interface ProductionVars {
+  REACT_APP_INFURA_API_KEY: string;
+  REACT_APP_PRICE_UPDATER_API_URL: string;
+  REACT_APP_PRICE_UPDATER_API_KEY: string;
+  REACT_APP_WALLETCONNECT_BRIDGE: string;
+}
+
+type ProductionEnv = ProductionLiteral & ProductionVars;
+
+interface DevelopmentLiteral {
+  REACT_APP_ENV: "development";
+}
+
+type DevelopmentVars = ProductionVars & {
+  REACT_APP_HERMEZ_API_URL: string;
+  REACT_APP_HERMEZ_CONTRACT_ADDRESS: string;
+  REACT_APP_WITHDRAWAL_DELAYER_CONTRACT_ADDRESS: string;
+  REACT_APP_BATCH_EXPLORER_URL: string;
+  REACT_APP_ETHERSCAN_URL: string;
+};
+
+type DevelopmentEnv = DevelopmentLiteral & DevelopmentVars;
+
+type Env = ProductionEnv | DevelopmentEnv;
+
+const productionLiteralParser = StrictSchema<ProductionLiteral>()(
+  z.object({
+    REACT_APP_ENV: z.literal("production"),
+  })
+);
+
+const productionVarsParser = StrictSchema<ProductionVars>()(
+  z.object({
+    REACT_APP_INFURA_API_KEY: z.string(),
+    REACT_APP_PRICE_UPDATER_API_URL: z.string(),
+    REACT_APP_PRICE_UPDATER_API_KEY: z.string(),
+    REACT_APP_WALLETCONNECT_BRIDGE: z.string(),
+  })
+);
+
+const productionEnvParser = StrictSchema<ProductionEnv>()(
+  productionLiteralParser.and(productionVarsParser)
+);
+
+const developmentLiteralParser = StrictSchema<DevelopmentLiteral>()(
+  z.object({
+    REACT_APP_ENV: z.literal("development"),
+  })
+);
+
+const developmentVarsParser = StrictSchema<DevelopmentVars>()(
+  productionVarsParser.and(
+    z.object({
+      REACT_APP_HERMEZ_API_URL: z.string(),
+      REACT_APP_HERMEZ_CONTRACT_ADDRESS: z.string(),
+      REACT_APP_WITHDRAWAL_DELAYER_CONTRACT_ADDRESS: z.string(),
+      REACT_APP_BATCH_EXPLORER_URL: z.string(),
+      REACT_APP_ETHERSCAN_URL: z.string(),
+    })
+  )
+);
+
+const developmentEnvParser = StrictSchema<DevelopmentEnv>()(
+  developmentLiteralParser.and(developmentVarsParser)
+);
+
+export function getEnv(): ZodEither<Env, ZodError<Env>> {
+  const parsedEnvLiteral = z
+    .union([productionLiteralParser, developmentLiteralParser])
+    .safeParse(process.env);
+
+  if (parsedEnvLiteral.success) {
```

Comment:
The `union` that you propose was also my initial approach but I had to change it because it has a big drawback in terms of error reporting: When the decoding of the union fails you get a list with the errors from all the members of the union.

Let's say you have the 5 vars from development commented (as usual in production). If you have a production variable wrong, then you get 6 errors. This was a no go IMO so I decided to first decode the env var `REACT_APP_ENV` to decide what decoder to use later and produce meaningful error messages.

> The interesting thing here is that you have the `env vars` wrong.

What is wrong? I'm not sure I follow you here.

Reply:
We agreed to keep the current implementation as it provides more accurate errors to the user.

Reply:
I tend to prefer simpler approaches, and Elias' one is easier to read. I think it's also enough to just notify that the environment variables are wrong.

Also +1 that production vars are actually shared vars, not specific to production.